threads/models.tsp

```diff
@@ -13,6 +13,54 @@ model CreateThreadRequest {
   /** A list of [messages](/docs/api-reference/messages) to start the thread with. */
   messages?: CreateMessageRequest[];
 
+  /**
+   * A set of resources that are made available to the assistant's tools in this thread. The
+   * resources are specific to the type of tool. For example, the `code_interpreter` tool requires
+   * a list of file IDs, while the `file_search` tool requires a list of vector store IDs.
+   */
+  tool_resources?: {
+    code_interpreter?: {
+      /**
+       * A list of [file](/docs/api-reference/files) IDs made available to the `code_interpreter`
+       * tool. There can be a maximum of 20 files associated with the tool.
+       */
+      @maxItems(20)
+      file_ids?: string[] = [];
+    };
+
+    file_search?: {
+      /**
+       * The [vector store](/docs/api-reference/vector-stores/object) attached to this thread.
+       * There can be a maximum of 1 vector store attached to the thread.
+       */
+      @maxItems(1)
+      vector_store_ids?: string[];
+
+      /**
+       * A helper to create a [vector store](/docs/api-reference/vector-stores/object) with
+       * file_ids and attach it to this thread. There can be a maximum of 1 vector store attached
+       * to the thread.
+       */
+      @maxItems(1)
+      vector_stores?: {
+        /**
+         * A list of [file](/docs/api-reference/files) IDs to add to the vector store. There can be
+         * a maximum of 10000 files in a vector store.
+         */
+        @maxItems(10000)
+        file_ids?: string[];
+
+        /**
+         * Set of 16 key-value pairs that can be attached to a vector store. This can be useful for
+         * storing additional information about the vector store in a structured format. Keys can
+         * be a maximum of 64 characters long and values can be a maxium of 512 characters long.
+         */
+        @extension("x-oaiTypeLabel", "map")
+        metadata?: Record<string>;
+      }[];
+    };
+  } | null;
```

Comment:
```suggestion
    file_search?: FileSearchVectorStores
  } | null;
 
 
 union FileSearchVectorStores {
     {
         /**
       * The [vector store](/docs/api-reference/vector-stores/object) attached to this thread.
       * There can be a maximum of 1 vector store attached to the thread.
       */
      @maxItems(1)
      vector_store_ids?: string[];
     },
     {
         /**
       * A helper to create a [vector store](/docs/api-reference/vector-stores/object) with
       * file_ids and attach it to this thread. There can be a maximum of 1 vector store attached
       * to the thread.
       */
      @maxItems(1)
      vector_stores?: {
        /**
         * A list of [file](/docs/api-reference/files) IDs to add to the vector store. There can be
         * a maximum of 10000 files in a vector store.
         */
        @maxItems(10000)
        file_ids?: string[];

        /**
         * Set of 16 key-value pairs that can be attached to a vector store. This can be useful for
         * storing additional information about the vector store in a structured format. Keys can
         * be a maximum of 64 characters long and values can be a maxium of 512 characters long.
         */
        @extension("x-oaiTypeLabel", "map")
        metadata?: Record<string>;
      }[];
     }
  }
```

Reply:
This change is to be able to express that either one or the other properties are required
![image](https://github.com/joseharriaga/openai-in-typespec/assets/21109913/67415d91-a02a-464c-af45-fe7b9056abe5)

Reply:
This one's going to be really interesting to try to accurately generalize... it's clear what the OpenAPI is trying to do ("one or the other") but I don't think it strictly specifies that intention as-is; one-of into distinct required lists doesn't force exclusion outside of required, so I think your TypeSpec does a better job of doing what was intended than they did!

Gotta think about this one, for sure.

Reply:
Upon looking at it a bit more, I think it's potentially even a bit more complicated than the union -- I suspect the `file_search` option is meant to remain open to new properties being added while still retaining the apparent "either/or" for IDs vs. creation helpers.

To do that, I added a bit of custom .tsp in the threads folder. I think it captures the rich semantics appropriately now, with the downside of adding a couple of extra visible types.

Let me know if that looks reasonable!